Add e2e package

errordeveloper opened the pull request:
This package contains bare minimum configuration required to run e2e tests.

As seen in https://github.com/linuxkit/linuxkit/pull/2668 earlier.

It still needs a little work (e.g. Job config would be plausible to add), but I wanted to open a PR before this gets lost.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "master" git@github.com:errordeveloper/linuxkit-kubernetes.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

ijc commented on pkg/e2e/Makefile:
```diff
@@ -0,0 +1,21 @@
+kube_release_artefacts=https://dl.k8s.io/$(kubernetes_version)/bin/linux/amd64
+
+.PHONY: build
+build: _cache/kubernetes-test.tar.gz _cache/kubectl
```
I'd really like to avoid adding any more `Makefile` wrappers for package builds at this point.

Are these things large enough to be worth caching or can we just get away with fetching them in the `Dockerfile`.

If they are big and we think caching is worth while then we should look into enhancing the linuxkit tool to express this in the `build.yml`, perhaps with a new subsection of `depends` alongside the docker image cache one I added to get rid of the last `Makefiles` here? If we are going down this path then probably it would be best to have a design in an issue on the linuxkit/linuxkit repo first.

Note that my preference would be to simply avoid the cache (i.e. just fetch in the `Dockerfile`) unless the artefacts are really so large as to be problematic.

ijc commented on pkg/e2e/build.yml:
```diff
@@ -0,0 +1,6 @@
+org: linuxkit
+image: kube-e2e-test
+network: true
+disable-content-trust: true
```
Lets setup trust for this repo before we merge this PR. I think everything in the linuxkit org really ought to be signed these days.

errordeveloper commented:
Ian, thanks for looking at this!

I do not like Makefile wrapper also, just checked it in as I had it
already. I thought of compiling it, like we do for kubelet package, but
then sharing the build logic and caching artefacts comes to mind...
If we do lean towards fetching binaries rather then building, I am happy to
think of a design for that, but personally I feel like we will build
Kubernetes code anyway, so we should just tackle that in a way that allows
us to reuse artefacts for different images. I certainly would like to be
able to use local checkout of k/k, or use form URL.

I am happy to copy build logic from kubelet for now, let me know what you
think. By the way, are we looking to eventually use buildkit for building
things from source? If so, how long until we might be able to use it here?

On Mon, 18 Dec 2017, 5:00 pm Ian Campbell, <notifications@github.com> wrote:

> I think you'll need to add a new job to the .circleci/config.yml,
> hopefully the pattern will be obvious, ask if not. (another reason to avoid
> Makefile wrappers is it avoids special cases here...)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/linuxkit/kubernetes/pull/35#issuecomment-352488437>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPWS3aaKgZ1-Of3elF0kGhk9G6P5E_pks5tBpo0gaJpZM4RFwq0>
> .
>

ijc commented:
It should be a relatively simply change to this PR to move the curl invocations into the Dockerfile nowm then we can get this merged (assuming the downloads are 10s to asmall hundred MB should be tolerable given we ought to be building this package reasonably infrequently) and consider whether we want to build from scratch or not (`kubelet` went through a similar trajectory with getting binaries first and switching to building later on).

WRT building, what is the source repo for this stuff?

ijc commented:
I missed replying to your final paragraph, #2587 has a bullet point for using buildkit but AFAIK noone is actively working on that so there is no timeline for when it might become possible.

errordeveloper commented:
> WRT building, what is the source repo for this stuff?

kubernetes/kubernetes

errordeveloper commented:
This now builds all binaries that are needed from source.

ijc commented on pkg/e2e/.gitignore:
```diff
@@ -0,0 +1 @@
+_cache
```
I don't think this exists in the current version so this whole file is redundant.

errordeveloper replied:
Indeed

ijc commented on pkg/e2e/Dockerfile:
```diff
@@ -0,0 +1,65 @@
+RUN mkdir -p /out/etc/apk && cp -r /etc/apk/* /out/etc/apk/
+#coreutils needed for du -B for disk image checks made by kubelet
+# example: $ du -s -B 1 /var/lib/kubelet/pods/...
+#          du: unrecognized option: B
```
Is this relevant in the context of building the tests?

errordeveloper replied:
No

ijc commented on pkg/e2e/README.md:
````diff
@@ -0,0 +1,27 @@
+```
+
+This will result in `linuxkit/kube-e2e-test:current` image that
+you can use. See `e2e.sh` for supported environment variables.
````
The section is out of date I think.

errordeveloper replied:
Yes

errordeveloper commented:
@ijc I've updated this, and included YAML in the top-level directory, as I didn't know where it maybe be better to put it. Inside the package it would affect the hash and that doesn't seem right.
I've also cleaned-up a few things, please take a look and let me know what you think. 

Running the e2e tests in CI would be a long job (I think it took 40min on my laptop last time I tried it), what do you think?
I certainly should add package build job to the CI, as discussed earlier.

ijc commented:
About to take a look, but came here to ask you to rebase onto #36 and to update the two comments I added and add a similar comment to the place you've added another Kube version definition.

> Running the e2e tests in CI would be a long job (I think it took 40min on my laptop last time I tried it), what do you think?
> I certainly should add package build job to the CI, as discussed earlier.

40 mins is certainly too long, the current set of tests takes ~20 mins in the worst case (all packages needing to be rebuilt) and can be much faster if packages don't need building.

We could consider adding a daily/nightly job which ran on master and ran these tests, but lets defer that until these basics are in.

ijc commented:
> included YAML in the top-level directory, as I didn't know where it maybe be better to put it. Inside the package it would affect the hash and that doesn't seem right.

Do you mean `test/kube-e2e-test-job.yaml`? Since it includes `linuxkit/kube-e2e-test:260c4467ab0a752b5369e73d4f2da13eb2da7bbf` I think it can't go in `pkg/e2e-test` even if you wanted to (since updating the hash would change the hash, rinse and repeat).

Actually that does highlight that this is not actually a "linuxkit package" (i.e. something you would feed to `linuxkit build` via a yml), it's an image intended to be run as a pod. It's not too bad to reuse `linuxkit pkg` to rebuild such a thing but perhaps it should be in `test/pkg` rather than `pkg` and then having `test/foo-job.yaml` which consumes it seems entirely reasonable.

Wherever it ends up you should include it in the `update-hashes` rune in the top-level `Makefile`.

What about also adding `yml/e2e-test.yml` similar to `yml/weave.yml` to pull in `kube-e2e-test-job.yaml` into `/etc/kubeadm/kube-system.init/99-e2e-test.yaml` so one can easily build an image which runs the tests automatically. Or are there issues like needing to wait after `kubeadm init` for things to stabilise before running the tests in order to get useful results?

ijc commented:
I think your CI issues are down to the thing which will be fixed by #39. Once we have the code in shape I can bash the retry button till it passes (or maybe #39 will get merged in time for you to rebase onto it)

errordeveloper commented:
> We could consider adding a daily/nightly job which ran on master and ran these tests, but lets defer that until these basics are in.

💯 👍

ijc commented on .circleci/config.yml:
```diff
@@ -330,6 +335,7 @@ workflows:
             - lint
             - pkg-kubelet
             - pkg-cri-containerd
+            - pkg-kube-e2e-test
```
You are missing a `linuxkit pkg push --nobuild` for this new package in the `push-pkgs-to-hub` step.

errordeveloper replied:
Well spotted!

ijc commented on pkg/kube-e2e-test/build.yml:
```diff
@@ -0,0 +1,6 @@
+org: linuxkit
+image: kube-e2e-test
+network: true
+disable-content-trust: true
```
I setup content trust for this repo so you can now drop this line.

errordeveloper replied:
Done.

ijc commented on scripts/run-e2e-test.sh:
```diff
@@ -0,0 +1,128 @@
+#!/bin/sh
+
+set -o errexit
+set -o pipefail
```
`pipefail` (at least) is a bash-ism so you need `#! /bin/bash` shebang. Not sure about `errexit` or `nounset`, the short names (`-e` and `-u`) are portable I think.

Given how awful it is to replicate `pipefail` in pure POSIX shell I think just using `/bin/bash` for the shebang is the right answer in this case.

errordeveloper replied:
Ah, I didn't realise this... It's okay, we aren't relying on `pipefail` here. I'll switch to `/bin/sh` and use `set -eu`, same goes for `pkg/kube-e2e-test/e2e.sh` (which currently has an explicit `set +o pipefail`).

ijc commented on scripts/run-e2e-test.sh:
```diff
@@ -0,0 +1,134 @@
+echo "$0: log save in ${PWD}/e2e.log, cleaning up the resources..."
+cleanup 2> /dev/null || true
+grep -q '^Test Suite Passed$' e2e.log && exit 0
+grep -q '^Test Suite Failed$' e2e.log && exit 1
```
If the log contains neither the pass nor fail messages that is a failure IMHO (perhaps the file is empty because pod logging is b0rked), I think you can just drop the `grep` from this last line and `exit 1` unconditionally.

errordeveloper commented:
I found an issue with the script, testing a fix right now.

errordeveloper commented:
@ijc I'm quite happy with the runner script now, PTAL.

errordeveloper commented:
> ```
> WARNING: The following packages cannot be authenticated!
>  libgdbm3 libgmp10 libnettle6 libhogweed4 libidn11 libffi6 libp11-kit0
>  libtasn1-6 libgnutls30 libkeyutils1 libkrb5support0 libk5crypto3 libkrb5-3
>  libgssapi-krb5-2 libunistring0 libsasl2-modules-db libsasl2-2 libldap-common
>  libldap-2.4-2 libnghttp2-14 libpsl5 librtmp1 libssh2-1 liberror-perl libxau6
>  libpopt0 krb5-locales less libbsd0 libncurses5 libedit2 libgpm2
>  openssh-client ca-certificates libsasl2-modules libxdmcp6 libxcb1
>  libx11-data libx11-6 libxext6 libxmuu1 patch rename xauth publicsuffix
> E: There were unauthenticated packages and -y was used without --allow-unauthenticated
> Exited with code 100
```
@ijc I never quite understood apt errors like this, any ideas?

ijc commented on .circleci/config.yml:
```diff
@@ -272,8 +274,9 @@ jobs:
-            linuxkit pkg push --nobuild pkg/cri-containerd
             linuxkit pkg push --nobuild pkg/kubelet
+            linuxkit pkg push --nobuild pkg/cri-containerd
+            linuxkit pkg push --nobuild pkg/kube-e2e-test
```
Why did you move cri-containerd below kubelet? They were alphabetic before. Suppose it isn't a big deal.

errordeveloper replied:
They were in different order in some places, so I was trying to do list them in the same order everywhere..

ijc replied:
I guess it was the other places which were buggy then ;-)

Lets not worry about it, if someone enthusiastic soul decides they want to sort them later (and add comments to keep them sorted) then they can.
